modules/globebrowsing/src/geojson/globegeometryfeature.cpp

```diff
@@ -0,0 +1,918 @@
+{
+    ghoul_assert(pass >= 0 && pass < 2, "Render pass variable out of accepted range");
+
+    float opacity = mainOpacity * _properties.opacity();
+    float fillOpacity = mainOpacity * _properties.fillOpacity();
+
+    const glm::dmat4 globeModelTransform = _globe.modelTransform();
+    const glm::dmat4 modelViewTransform =
+        renderData.camera.combinedViewMatrix() * globeModelTransform;
+
+    const glm::mat3 normalTransform = glm::mat3(
+        glm::transpose(glm::inverse(modelViewTransform))
+    );
+
+    const glm::dmat4 projectionTransform = renderData.camera.projectionMatrix();
+
+#ifndef __APPLE__
+    glLineWidth(_properties.lineWidth());
+#else
+    glLineWidth(1.f);
+#endif
+
+    for (const RenderFeature& r : _renderFeatures) {
+        if (r.isExtrusionFeature && !_properties.extrude()) {
+            continue;
+        }
+
+        bool shouldRenderTwice = r.type == RenderType::Polygon &&
+            fillOpacity < 1.0 && _properties.extrude();
+
+        if (pass > 0 && !shouldRenderTwice) {
+            continue;
+        }
+
+        ghoul::opengl::ProgramObject* shader = (r.type == RenderType::Points) ?
+            _pointsProgram : _linesAndPolygonsProgram;
+
+        shader->activate();
+        shader->setUniform("modelTransform", globeModelTransform);
+        shader->setUniform("viewTransform", renderData.camera.combinedViewMatrix());
+        shader->setUniform("projectionTransform", projectionTransform);
+
+        shader->setUniform("heightOffset", _offsets.z);
+        shader->setUniform("useHeightMapData", useHeightMap());
+
+        if (shader == _linesAndPolygonsProgram) {
+            shader->setUniform("normalTransform", normalTransform);
+            shader->setUniform("nLightSources", lightSourceData.nLightSources);
+            shader->setUniform("lightIntensities", lightSourceData.intensitiesBuffer);
+            shader->setUniform(
+                "lightDirectionsViewSpace",
+                lightSourceData.directionsViewSpaceBuffer
+            );
+        }
+
+        glBindVertexArray(r.vaoId);
+
+        switch (r.type) {
+            case RenderType::Lines:
+                shader->setUniform(
+                    "opacity",
+                    r.isExtrusionFeature ? fillOpacity : opacity
+                );
+                renderLines(r);
+                break;
+            case RenderType::Points:
+                shader->setUniform("opacity", opacity);
+                renderPoints(r, renderData, pointRenderMode);
+                break;
+            case RenderType::Polygon: {
+                shader->setUniform("opacity", fillOpacity);
+                renderPolygons(r, shouldRenderTwice, pass);
+                break;
+            }
+            default:
+                throw ghoul::MissingCaseException();
+                break;
+        }
+
+        shader->deactivate();
+    }
+
+    glBindVertexArray(0);
+
+    // Reset after every geometry
```

Comment:
if you remember why, maybe a short message why it is necessary to reset the state after every single geometry?